stdlib/http/src/main/java/org/ballerinalang/net/http/compiler/ResourceSignatureValidator.java

```diff
@@ -70,27 +75,139 @@ public static void validateAnnotation(FunctionNode resourceNode, DiagnosticLog d
+        List<? extends SimpleVariableNode> signatureParams = resourceNode.getParameters().subList(
+                COMPULSORY_PARAM_COUNT, resourceNode.getParameters().size());
+        if (!signatureParams.stream().allMatch(signatureParam -> paramSegments.stream()
+                .anyMatch(parameter -> signatureParam.getName().getValue().equals(parameter)))) {
+            dlog.logDiagnostic(Diagnostic.Kind.ERROR, resourceNode.getPosition(),
+                               "Invalid parameter(s) in the resource signature");
+        }
+    }
+
+    private static void validateWebSocketUpgrade(FunctionNode resourceNode, DiagnosticLog dlog,
+                                                 List<BLangRecordLiteral.BLangRecordKeyValue> annVals,
+                                                 List<String> paramSegments,
+                                                 BLangRecordLiteral.BLangRecordKeyValue keyValue) {
+        if (annVals.size() > 1) {
+            dlog.logDiagnostic(Diagnostic.Kind.ERROR, resourceNode.getPosition(),
+                               "Invalid configurations for WebSocket upgrade resource");
+            return;
+        }
+        List<BLangRecordLiteral.BLangRecordKeyValue> upgradeFields =
+                ((BLangRecordLiteral) keyValue.valueExpr).keyValuePairs;
+        if (upgradeFields.isEmpty()) {
+            dlog.logDiagnostic(Diagnostic.Kind.ERROR, resourceNode.getPosition(),
+                               "An upgradeService need to be specified for the WebSocket upgrade " +
+                                       "resource");
+            return;
+        }
+        if (upgradeFields.size() == 1 && !(getAnnotationFieldKey(upgradeFields.get(0)).equals("upgradeService"))) {
+            dlog.logDiagnostic(Diagnostic.Kind.ERROR, resourceNode.getPosition(),
+                               "An upgradeService need to be specified for the WebSocket upgrade " +
+                                       "resource");
+            return;
+        }
+        // WebSocket upgrade path validation
+        for (BLangRecordLiteral.BLangRecordKeyValue upgradeField : upgradeFields) {
+            if (getAnnotationFieldKey(upgradeField).equals(ANN_WEBSOCKET_ATTR_UPGRADE_PATH)) {
+                validateResourcePath(dlog, paramSegments, upgradeField);
+            }
+        }
+    }
+
+    private static void validateResourcePath(DiagnosticLog dlog, List<String> paramSegments,
+                                             BLangRecordLiteral.BLangRecordKeyValue keyValue) {
+        DiagnosticPos position = keyValue.getValue().getPosition();
+        String[] segments = keyValue.getValue().toString().split("/");
+        for (String segment : segments) {
+            validatePathSegment(segment, position, dlog, paramSegments);
+        }
+    }
+
+    private static void validatePathSegment(String segment, DiagnosticPos pos, DiagnosticLog dlog,
```

Comment:
What if some user writes something like /foo/{}/bar ? does it properly get validated ?

Reply:
It gives compilation error saying `Illegal empty expression in resource path config`

Related test case : https://github.com/ballerina-platform/ballerina-lang/pull/15979/files#diff-d0ef7231cf640ec02b8786e5ce779dfdR50